### ISSUE TITLE
null checked variables in getter's closure

### ISSUE THREAD

**Issue by vmit** [microsoft/TypeScript#18243]
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)

**Code**
```ts
export function foo(z: string | null) {
    const zConst = z;
    if (zConst !== null) {
        return {
            get bar() {
                return zConst.length;
            }
        }
    }
}
```

**Expected behavior:**
The code is compiled with "strictNullChecks=true" option as **zConst** is null-checked and could not be changed.
**Actual behavior:**
TS is complaining about **zConst** being possibly null.

The problem seems is relevant to getters only since with the same **zConst** in a simple function scope it is fine.

**Comment by FoxShaunR** [comment]
This isn't limited to null checks. Ran into the same issue with a discriminating union. [Example here](https://www.typescriptlang.org/play?#code/KYOwrgtgBAMg9gQwCYGUAuC3CgbwFBRQDqCAlmqSAOYA0BsiSwSdhAogE4dwd0C+eQZSwcAZggDG2eMjYA3UGk7cOuegGcMWAFwNk6TMAB0yngG56wLj12mOFgXmFXxUvUnmKZTJGsKbDXW8DLCNvZgtCOQQAGzBgXTAQAGsQOAB3EAchEBFXaUZPXJJySio-KACddxDjEopqbLw0AE8ABwLZBVyoAF53IqVrVQAfAe60cN8x70H6sos8CRiEdXUoABVgTQBhFbWKwiQ4DYALMvUACmAJoMKJgEpDwhfSUShriaMq7F6-mq0ximT3wLzB4IA9BCoDdFFAJAgQFBTggFFBonFsKcrMB6ODwRI4CB1HAYsYYnAqJ9FEYMfEHpF8QSiZooFRgGgRH1nkzwey0FAACSkCBtGKkCTkS4PRIpNKZHm8plQmETKAQUhUU4CtIClFoulYnF4pW8jgcsAcJGw3K02LxRmmsECJ1QPiO8Eul4CARAA)

**Comment by jakebailey** [comment]
Just checking old issues; this one was fixed by #45020 (#45006 was the same as this issue).
